src/ClientImpl.cpp

```diff
@@ -664,8 +870,14 @@ InvocationResponse ClientImpl::invoke(Procedure &proc) throw (voltdb::Exception,
     if (evbuffer_add(evbuf, sbb.bytes(), static_cast<size_t>(sbb.remaining()))) {
         throw voltdb::LibEventException();
     }
+    timeval tv, expirationTime;
+    int status = gettimeofday(&tv, NULL);
+    assert(status == 0);
+    expirationTime.tv_sec = tv.tv_sec + m_queryExpirationTime.tv_sec;
+    expirationTime.tv_usec = tv.tv_usec + m_queryExpirationTime.tv_usec;
+    boost::shared_ptr<CallBackBookeeping> cb (new CallBackBookeeping(callback, expirationTime));
     m_outstandingRequests++;
-    (*m_callbacks[bev])[clientData] = callback;
+    (*m_callbacks[bev])[clientData] = cb;
 
     if (event_base_dispatch(m_base) == -1) {
```

Comment:
If dispatch fails, it will generate exception which will destrict the call back bev map and destruct all callbacks automatically (callback resource is wrapped in shared ptr)